### ISSUE TITLE
Possibility of dual-licensing?

### ISSUE THREAD

**Issue by thomasfsteeples** [sheredom/utf8.h#100]
utf8.h is currently published under the [Unlicense](https://unlicense.org/), putting its work in the public domain. This is great, but there are [open questions](
 https://softwareengineering.stackexchange.com/questions/147111/what-is-wrong-with-the-unlicense) as to whether this is valid in all jurisdictions (Germany being the most famous example).

As such, would you be at all willing to consider dual-licensing this software under the Unlicense and another "fallback" license? The [CC0 license](https://creativecommons.org/share-your-work/public-domain/cc0/) is another public domain license with a clause for what should happen when the terms of the license are deemed invalid under local law. Alternatively, there exist other minimal OSI approved licenses (such as the MIT license, the ISC license and the BSD licenses) which are permissive. These typically require attribution from the user, but if the software were dual-licensed, it would be entirely their choice which license they want to use.

Absolutely no worries if this is too big an ask, just really want to be able to use this software in a more legally-watertight way.

**Comment by sheredom** [comment]
I did think about CC0 - but it likewise hasn't be tried in a court to prove it works. I have a personal distaste for dual licensing just because it adds words (stupid reason I know) but if this was the reason stopping you using the library I might consider it. I know the other popular libraries https://github.com/nothings/stb are dual licensed.

**Comment by thomasfsteeples** [comment]
> I have a personal distaste for dual licensing just because it adds words (stupid reason I know) but if this was the reason stopping you using the library I might consider it.

So this is currently stopping me from using the library for one project, but I certainly use utf8.h for other things. Entirely up to you, but I feel dual-licensing with MIT, say, offers flexibility to users, without the project really losing anything. Equally, I completely respect the choice to public domain the code, and to _not_ dual license it. Ultimately, up to you, and happy for you to close the issue with whatever you decide.

**Comment by untyper** [comment]
If dual licensing is necessary, then I think it's a much better idea to use the more permissive license's of Zlib or Boost rather than MIT.

**Comment by sheredom** [comment]
This is the problem with dual licensing in general tbh - no-one can agree on what is best. In reality the 'safest' dual license is Apache 2.0 for the patent waiver fun (yay 😢) - but it'd be more in the spirit of my contributions to dual license between unlicense and CC0.

**Comment by crt333** [comment]
Would the MIT No Attribution license (MIT-0) work?

**Comment by Febbe** [comment]
@thomasfsteeples That license would be valid even when parts of it are invalid. It's just not possible to transfer or deny your copyright completely. But it will still be interpreted from courts as a license grant.

**Comment by sheredom** [comment]
I've been dwelling on this and I've decided to leave the license as is. Having a single license makes it easier on me (cognitively to know what I am signing up for) and on my existing many customers in that they already know what they have signed up for.

Given the unliensce is an attempt to explicitly say the work is public domain I will explicitly note you can take the code and relicense it if you choose. No skin off my ass!

**Comment by darealshinji** [comment]
> Given the unliensce is an attempt to explicitly say the work is public domain I will explicitly note you can take the code and relicense it if you choose.

I can't find such a note in the license header. Can you add this to the header of the source file?

Personally I think you should still reconsider relicensing to a less problematic permissive license like MIT/BSD/zlib. Right now as a German I cannot use this in commercial software because it's not possible to relinquish your rights on your own work. The Unlicense is void and according to German law this software is the same as if you had written "Copyright Neil Henning. All rights reserved." into the header file. This means you could technically demand money if I or someone else in Germany uses this in a commercial project.
And if I read the discussion on Stackexchange the validity in some other jurisdictions is unknown or at least questionable. It's not that I don't like the idea behind the Unlicense but I will have to always think twice if and where I use such code.

The issue with the Unlicense is that it doesn't explicitly mention what to do in case the first paragraph is considered invalid.
I think the idea behind a permissive software license is that you still formally claim copyright but make it so easy to fulfill the license conditions that it's effectively the same as PD code.

**Comment by Febbe** [comment]
> > Given the unliensce is an attempt to explicitly say the work is public domain I will explicitly note you can take the code and relicense it if you choose.
> 
> I can't find such a note in the license header. Can you add this to the header of the source file?
> 
> Personally I think you should still reconsider relicensing to a less problematic permissive license like MIT/BSD/zlib. Right now as a German I cannot use this in commercial software because it's not possible to relinquish your rights on your own work. The Unlicense is void and according to German law this software is the same as if you had written "Copyright Neil Henning. All rights reserved." into the header file. This means you could technically demand money if I or someone else in Germany uses this in a commercial project.
> And if I read the discussion on Stackexchange the validity in some other jurisdictions is unknown or at least questionable. It's not that I don't like the idea behind u but I will have to always think twice if and where I use such code.
> 
> The issue with the Unlicense is that it doesn't explicitly mention what to do in case the first paragraph is considered invalid.
> I think the idea behind a permissive software license is that you still formally claim copyright but make it so easy to fulfill the license conditions that it's effectively the same as PD code.
> 

No, there is no problem with that licence, since the license is splittable. And if you are unsure, copy the statement above and print it. It's also a license grant (Asking the owner and getting the permission is a license grant even if he would gave you this permission on toilet paper)

In any case, the unlicense license explicitly states:
> Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means. 

This sentence is a fully valid license grant.
So you can do exactly this without any fear. 

The only questionable part is the following, it basically denies all copyright, which is indeed not possible (§ 29 Abs. 1 UrhG).
This part can, but is not required to be interpreted as a full license grant for everything including to relicense something.

But in any case, if there would be a dispute, the license falls back to the above statement (§ 31 abs. 5 UrhG and § 139 BGB). There is no need for a "Salvatortische Klausel" in German law, since it's exactly what § 139 BGB already states.

So if you are unsure, you can still assume, that you have full use rights, beside of relicensing.
Basically using MIT as license would add nothing.

**Comment by darealshinji** [comment]
Thanks for the clarification. However, I think there's a reason why companies like Google avoid licenses such as Unlicense. I've switched to GLib's utf8 functions anyway.
